### ISSUE TITLE
no-unused-vars removal option

### ISSUE THREAD

**Issue by merlinpatt** [eslint/eslint#11668]
**The version of ESLint you are using.**
5.16.0

**The problem you want to solve.**
Easily remove unused variables and especially imports.

I know that this was shot down in #7619 but that was to auto remove them on any `--fix`. This would only do so with an extra flag.

**Your take on the correct solution to problem.**
Have a command line (`--remove-unused`) and config (`removeUnused`) option to remove unused variables and imports for you.

**Are you willing to submit a pull request to implement this change?**
Possibly, if I can learn how create the flag and code needed. I haven't worked on a JS command line tool before.

**Comment by platinumazure** [comment]
Hi @merlinpatt, thanks for the issue.

I don't think a command line or configuration option makes sense here, as this is specific to one rule. As for handling the desired use case, it would be pretty easy to write a tool that runs the rule and applies its own changes- or you could write a custom ESLint rule using [`eslint-rule-composer`](https://github.com/not-an-aardvark/eslint-rule-composer) which iterates over the reported errors and adds an autofix.

In other words, this is already possible for users to do with only a little extra work (which could then be shared with others), so I don't see much value in adding this to ESLint core.

**Comment by kaicataldo** [comment]
I also don't think we should be adding additional CLI flags for specific rules (that would also be a core change and require an [RFC](https://github.com/eslint/rfcs).

I'm not convinced it's desirable to auto remove imports in particular - there are valid cases where a module might be imported and then not referenced by name anywhere else in the file (example [here](https://github.com/yannickcr/eslint-plugin-react/blob/8ef86c53003d78c88416a35e1e627914fcfabdf4/docs/rules/jsx-uses-react.md))

**Comment by merlinpatt** [comment]
Auto removal would only remove things that aren't used. React is a special case and as noted, it does get used, just not referenced by name.

As for writing my own tool, or composing one, can you explain how to do that? It may be easy to you but it's not easy to me.

**Comment by kaicataldo** [comment]
> Auto removal would only remove things that aren't used. React is a special case and as noted, it does get used, just not referenced by name.

What I'm unclear on is how you know if something is used or not. The React case might work with the above rule enabled, but what about other potential cases? What if something is used and removing it breaks the actual execution of the code (for instance, side effects from evaluating an imported module)? 

I share the same concerns as https://github.com/eslint/eslint/issues/7619#issuecomment-261541901 and https://github.com/eslint/eslint/issues/7619#issuecomment-261740255.

I'm not against the idea if it can be done safely (i.e. guaranteed to not break code execution!), but I'm not sure how we could ensure that given the limits of static analysis.

**Comment by merlinpatt** [comment]
You could add a list of packages that you don't want removed. Most packages are imported and used.

The reason of "people may use it later" doesn't hold up. If that's the case, they either don't use the autofix or they ignore particular files.

**Comment by kaicataldo** [comment]
> You could add a list of packages that you don't want removed. 

We could add a configuration option of a list of packages that are not allowed to be removed, but I personally don't know how we could ensure that the case @not-an-aardvark mentions [here](https://github.com/eslint/eslint/issues/7619#issuecomment-261740255) does not break anyone's code. One of the guiding principles of how ESLint approaches auto fixers is that we don't apply fixes that can potentially break code execution. That means that we need to err on the side of caution. In this case, a white list of things that can be removed seems like a much safer option, but then that mostly defeats the purpose of this enhancement, since you'd have to keep a list of all the modules in your project (yikes). All that being said, if you have a solution to this problem, we'd love to hear it.

> The reason of "people may use it later" doesn't hold up.

I believe you're referring to @mysticatea's comment [here](https://github.com/eslint/eslint/issues/7619#issuecomment-261541901). I think he is referring to the case where people run the autofixer on save in their text editor (please correct me if I'm wrong, @mysticatea). With this autofixing behavior, the import statement would be removed on save if you typed it out first and did not use it anywhere. Mind explaining why you believe this doesn't hold up?

I understand the desire for this feature (and how it could potentially work flawlessly for your specific use case), but we do have to think about all users and potential use cases (which are many!) when we evaluate proposals.

**Comment by g-plane** [comment]
@merlinpatt Additionally, if you really want this feature, you can open an issue here: https://github.com/aladdin-add/eslint-plugin-autofix .

**Comment by corysimmons** [comment]
What's the rationale behind not wanting to add a CLI flag or config for this?

CLI: `$ eslint --keep-libs="react,foo-bar"`

or

```js
"eslintConfig": {
  "keepLibs": ["react", "foo-bar"]
}
```

Also, can't you use Babel or something to include specific libs without them having to be exposed in the project? I know NextJS does this so you don't have to type `import React from 'react'` everywhere 😩

**Comment by g-plane** [comment]
@corysimmons Are you using React and JSX?

**Comment by corysimmons** [comment]
@g-plane Yes, I'm using NextJS and it auto-imports that stuff behind the scenes so I don't actually have to have `import React from 'react'` at the top of any of my files.

**Comment by g-plane** [comment]
However, I don't understand your use cases.

**Comment by corysimmons** [comment]
https://twitter.com/john_papa/status/991111269674569733?lang=en but as a CLI bin so I can add `"precommit": "eslint-and-sort-imports"` instead of having to use VSCode.

If I have a project with 500 files, I currently have to open each one and save it in order to sort imports.

If it is a CLI, then I can do them all with a glob pattern.

**Comment by g-plane** [comment]
If you just want to sort imports, please check out eslint-plugin-import.

**Comment by kaicataldo** [comment]
> What's the rationale behind not wanting to add a CLI flag or config for this?

@corysimmons Please the conversation above - I believe we address this in our comments. Happy to clarify if you have further questions. 

The reason I’m currently hesitant to include this behavior is because there are cases where this could break the actual execution of the code. If we can find a solution that guarantees this doesn’t happen, I think it would be a really nice feature to have (though I would expect it to be added to a rule and not added as a CLI flag).

VS Code only has to be concerned with a very specific use case (running on save in a text editor), while ESLint can be run on code written for any number of environments and execution contexts, so we have to be very careful when considering new features like this.

**Comment by g-plane** [comment]
Maybe this issue can be closed, because ESLint core doesn't intend to provide auto fixing for removing unused variables and imports, and there is an existing third-party ESLint plugin (eslint-plugin-autofix) which has implemented this feature.
